### ISSUE TITLE
Speed up international-aware JavaScript string functions

### ISSUE THREAD

**Issue by achlipala** [urweb/urweb#150]
Thanks to @fabriceleal for the internationalization in #146.  It has been pointed out that some of the JavaScript code is quite verbose.  I agree that this logic should be there, to maintain identical behavior with ICU on the server side.  However, I think the code can be rewritten to be much faster.  It is especially unpleasant to think of all-ASCII applications paying this cost on every call to such an operation.  @fabriceleal, are you up to attempting some fix?

You already used an automatic code generator to produce the sequential `if` code we see now.  Could you modify your generator to do a binary search through the number space or a bit-by-bit division of the value space?  Either one should lead to tree-shaped code (running time logarithmic in number of values) instead of sequential code (running time linear in number of values, unless JavaScript JITs are getting super-clever).

**Comment by fabriceleal** [comment]
I can give it a try, I did a quick change on the code generator to check for sequences `mod 2 = n` and found a decent amount of cases in islower. So I think I'll try to combine this with the binary search method and try to come up with something.

**Comment by fabriceleal** [comment]
What do you think about this approach: embed an array with the codepoints for which a function returns true (https://github.com/fabriceleal/urweb/commit/ea370b4e4b531206a8a9889e63309d3f7ff6d9c6#diff-867e8dfbbc36419eefc0dfdf9db32883R48), and using binary search (https://github.com/fabriceleal/urweb/commit/ea370b4e4b531206a8a9889e63309d3f7ff6d9c6#diff-867e8dfbbc36419eefc0dfdf9db32883R32) to seek the value in said array?

**Comment by achlipala** [comment]
Shouldn't my third suggestion be faster, but with essentially the same amount of code?  That suggestion was to initialize a JavaScript object with fields named after exactly those numbers that are in the character class.  Then it's just one field access!

**Comment by fabriceleal** [comment]
Not sure, we do need to initialize the hashtable for every single page, the same as the array... but I can make it happen. By the way, would you prefer a single hashtable with every property, or a hashtable for each function?

**Comment by achlipala** [comment]
I agree that initialization would look roughly the same, but then access is constant-time instead of logarithmic-time, right, with my suggestion?  And interesting question about hashtable per property vs. function.  I'm not really sure, but it probably doesn't make a huge difference.

**Comment by fabriceleal** [comment]
Alrighty, here it is: https://github.com/fabriceleal/urweb/commit/b0db19095cad90b30281b5bb105ac023c57b8cb2#diff-867e8dfbbc36419eefc0dfdf9db32883R32

I ran my exhaustive test that compares the results of islower server vs client side, and the difference seems negligible to me, although i havent compared load times specifically. My very unscientific test yielded

64.669s for a single full test run of the hashtable approach
64.508s ... with the array + binary search approach
65.050s ... with the "expanded code" binary search approach

If everything's ok with you I'll migrate the rest to the hashtable approach and submit a PR

**Comment by achlipala** [comment]
Yeah, sounds like a good plan; thanks.  I ask out of curiosity, did you compare performance against the original linear-test approach, to check that I wasn't being silly asking for a refactoring?

**Comment by fabriceleal** [comment]
I just assumed it would be the worst of all; I ran it 4 times and got 68.144s, 67.887s, 68.608.s and 67.860s
